app/include/drivers/behavior.h

```diff
@@ -51,23 +113,57 @@ __subsystem struct behavior_driver_api {
     behavior_keymap_binding_callback_t binding_released;
     behavior_sensor_keymap_binding_accept_data_callback_t sensor_binding_accept_data;
     behavior_sensor_keymap_binding_process_callback_t sensor_binding_process;
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+    behavior_get_parameter_domains_t get_parameter_domains;
+    const struct behavior_parameter_metadata_custom *custom_parameters;
+    const enum behavior_parameter_standard_domain param1_standard_domain;
+    const enum behavior_parameter_standard_domain param2_standard_domain;
+#endif // IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
 };
 /**
  * @endcond
  */
 
 struct zmk_behavior_ref {
     const struct device *device;
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+    const char *friendly_name;
+#endif // IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
 };
 
+#define ZMK_BEHAVIOR_REF_DT_NAME(node_id) _CONCAT(zmk_behavior_, DEVICE_DT_NAME_GET(node_id))
+
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+
+#define ZMK_BEHAVIOR_REF_INITIALIZER(_dev, _friendly_name)                                         \
+    { .device = _dev, .friendly_name = _friendly_name, }
+
+#define ZMK_BEHAVIOR_REF_DEFINE(name, ...)                                                         \
+    static const STRUCT_SECTION_ITERABLE(zmk_behavior_ref, name) =                                 \
+        ZMK_BEHAVIOR_REF_INITIALIZER(__VA_ARGS__)
+
+#define ZMK_BEHAVIOR_REF_DT_DEFINE(node_id)                                                        \
+    ZMK_BEHAVIOR_REF_DEFINE(ZMK_BEHAVIOR_REF_DT_NAME(node_id), DEVICE_DT_GET(node_id),             \
+                            DT_PROP_OR(node_id, friendly_name, DEVICE_DT_NAME(node_id)))
+
+#else
+
+#define ZMK_BEHAVIOR_REF_INITIALIZER(_dev)                                                         \
+    { .device = _dev, }
+
+#define ZMK_BEHAVIOR_REF_DEFINE(name, ...)                                                         \
```

Comment:
Does the compiler complain about unused macro parameters? If not, you could replace the variable args with an explicit parameter list and then just pass in everything but ignore the unneeded parameters whenever metadata is disabled. As this is currently written, it's quite difficult to understand what parameters this takes in either case.

Also, an idea for reducing the number of macros that need alternate implementations:

```c
struct zmk_behavior_metadata {
    // Making this still defined but empty when metadata is disabled so we can write
    // ".metadata = ..." and have it still compile when metadata is disabled.
#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
    const char *name;
    // any additional fields needed later go here...
#endif
};

#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
#define ZMK_BEHAVIOR_METADATA_INITIALIZER(name) \
    { .name = name }
#else
#define ZMK_BEHAVIOR_METADATA_INITIALIZER(name) {}
#endif    

struct zmk_behavior_ref {
    const struct device *device;
    struct zmk_behavior_metadata metadata;
};

#define ZMK_BEHAVIOR_REF_INITIALIZER(dev, name) \
    { .device = dev, .metadata = ZMK_BEHAVIOR_METADATA_INITIALIZER(name) }
```

(Test for that idea: https://godbolt.org/z/GzaeT9dYh)

Reply:
Yeah, I dig it. Took a slightly different approach though and just passed the node_id into the macro, in case more gets added later, the `ZMK_BEHAVIOR_METADATA_INITIALIZER` can handle that "internally".

Thoughts on the updated take?